src/Titanium.Web.Proxy/Network/CertificateManager.cs

```diff
@@ -472,61 +436,69 @@ internal X509Certificate2 CreateCertificate(string certificateName, bool isRootC
         internal async Task<X509Certificate2> CreateCertificateAsync(string certificateName)
         {
             // check in cache first
-            if (certificateCache.TryGetValue(certificateName, out var cached))
+            var item = cachedCertificates.GetOrAdd(certificateName, _ =>
             {
-                cached.LastAccess = DateTime.Now;
-                return cached.Certificate;
-            }
+                var cached = new CachedCertificate();
+                cached.CreationTask = Task.Run(() =>
+                {
+                    var certificate = CreateCertificate(certificateName, false);
+
+                    // see http://www.albahari.com/threading/part4.aspx for the explanation
+                    // why Thread.MemoryBarrier is used here and below
+                    cached.Certificate = certificate;
+                    Thread.MemoryBarrier();
+                    cached.CreationTask = null;
```

Comment:
Why are we setting the Task to null.

Reply:
This is the task that creates a certificate. When it finishes, it assigns Certificate property to the generated certificate and cleans CreationTask property to allow the Task memory to be released by GC.